[allocator.requirements.general] Remove redundant template syntax

AlisdairM opened the pull request:
In the Cpp17Allocator requirements, all identifiers name specific types, even if they are template instantiations.  Hence, all use of `typename`, `.template`, and `::template` is redundant.  Not in the sense of we are awaiting LWG to review P2150 for policy, but in the sense these have been redundant since C++11, and were outright ill-formed prior to that.

The majority of the wording throughout this subclause already (correctly) avoids use of these redundant keywords.  This PR brings the title rows into agreement with the body text, and fixes one outstanding redundant use of `::template`.

JohelEGP commented:
I think those `typename` are blessed now:
> [(3.7)](https://eel.is/c++draft/library#structure.specifications-3.7)
Result: for a [typename-specifier](https://eel.is/c++draft/temp.res.general#nt:typename-specifier), a description of the named type; for an [expression](https://eel.is/c++draft/expr.comma#nt:expression), a description of the type of the expression; the expression is an lvalue if the type is an lvalue reference type, an xvalue if the type is an rvalue reference type, and a prvalue otherwise[.](https://eel.is/c++draft/library#structure.specifications-3.7.sentence-1)

jwakely reviewed (commented):
These aren't code, they're descriptions of members. As such, I think it's useful to show that these are types. It's how you'd write it in a `requires-expression` too:

```c++
requires { typename X::pointer; }
```

I don't think this change would be an improvement.

tkoeppe commented on source/lib-intro.tex:
```diff
@@ -2133,7 +2133,7 @@
 \ensures
 For all \tcode{U} (including \tcode{T}),
-\tcode{Y::template rebind<T>::other} is \tcode{X}.
+\tcode{Y::rebind<T>::other} is \tcode{X}.
```
@jwakely Are there any other cases of `::template` that could be removed from the draft?

jwakely replied:
The Note 2 on [container.reqmts] p64 is unclear, but as a note I don't think it matters either way. I can't find any others.

tkoeppe replied:
OK, thanks! @AlisdairM, could you retool this PR please to only contain this one change, and revert all the others?

tkoeppe replied:
@AlisdairM Ping?

AlisdairM commented:
Better late than never --- finally rebased and pushed just the one change, as requested by @jwakely.  Rebased, and updated the commit message to match.
